### PR TITLE
Fix bugs in C API and refactor tests

### PR DESCRIPTION
**Pull request by klecki**
## Description
- [x] **Bug fix** (*non-breaking change which fixes an issue*)
- [ ] **New feature** (*non-breaking change which adds functionality*)
- [ ] **Breaking change** (*fix or feature that would cause existing functionality to not work as expected*)
- [x] **Refactoring** (*Redesign of existing code that doesn't affect functionality*)
- [ ] **Other** (*e.g. Documentation, Tests, Configuration*)

### What happened in this PR
Fix `daliDeserializeDefault` - the batch size
map was not allocated and when the pipeline
handle was correctly cleaned, it tried to
deallocate memory under invalid pointer.

Add missing `daliDeletePipeline` in all C API tests
so the tests no longer leak the created pipeline.

Some tests (daliOutputCopySamples) used to fill the
prefetch queue and schedule additional runs, 
whilst accessing only one output iteration. 
The not-freed pipeline could be actively using the memory
when the memory resources were cleared on shutdown
and the whole test process could crash during shutdown.
Remove the additional unused iteration in such case.

Rework tests to not use access to underlying
contiguous buffer of TensorList.

#### Additional information
- Affected modules and functionalities:
C API

- Key points relevant for the review:

### Tests
- [x] Existing tests apply
- [ ] New tests added
  - [ ] Python tests
  - [x] GTests
  - [ ] Benchmark
  - [ ] Other
- [ ] N/A

### Documentation
- [x] Existing documentation applies
- [ ] Documentation updated
  - [ ] Docstring
  - [ ] Doxygen
  - [ ] RST
  - [ ] Jupyter
  - [ ] Other
- [ ] N/A

#### Requirements
- [ ] Implements new requirements
- [ ] Affects existing requirements
- [x] N/A

**REQ IDs**: N/A

**JIRA TASK**: N/A

### CODE REVIEW

**Comment by klecki** [comment]
!build

File: dali/c_api/c_api_test.cc [hunk 1]
```diff
@@ -704,26 +740,32 @@ void TestForceFlagRun(bool ext_src_no_copy, unsigned int flag_to_test) {
...
                      prefetch_queue_depth, false);
 
-  std::vector<TensorList<TypeParam>> data;
-  data.resize(prefetch_queue_depth);
+  std::vector<std::shared_ptr<uint8_t>> data;
+  data.reserve(prefetch_queue_depth);
+  for (int i = 0; i < prefetch_queue_depth; i++) {
+    data.push_back(AllocBuffer<TypeParam>(num_elems * sizeof(uint8_t), false));
+  }
+  std::vector<TensorList<TypeParam>> input_wrapper(3);
```
Comment (JanuszL) [inline dali/c_api/c_api_test.cc:748]: Maybe it is better to use TensorVector unless TensorList is the data structure we are reworking towards?

Comment (klecki) [reply dali/c_api/c_api_test.cc:748]: Hmm, it should be 
```suggestion
  std::vector<TensorList<TypeParam>> input_wrapper(prefetch_queue_depth);
```
but other then that I just kept the TensorList. What we will use is up to be decided, but I've heard voices to keep the TensorList to match what we have in Python. But we will see.

**Review by JanuszL — APPROVED** [review]

File: dali/c_api/c_api.cc [hunk 2]
```diff
@@ -195,9 +195,11 @@ void daliDeserializeDefault(daliPipelineHandle *pipe_handle, const char *seriali
     stream = dali::CUDAStream::Create(true);
   }
   auto ws = std::make_unique<dali::DeviceWorkspace>();
+  auto bs_map = std::make_unique<batch_size_map_t>();
   pipe_handle->ws = ws.release();
   pipe_handle->copy_stream = stream.release();
   pipe_handle->pipe = pipeline.release();
+  pipe_handle->batch_sizes_map = bs_map.release();
```
Comment (mzient) [inline dali/c_api/c_api.cc:202]:
```suggestion
  pipe_handle->batch_size_map = bs_map.release();
```

File: dali/c_api/c_api.cc [hunk 3]
```diff
@@ -177,11 +177,11 @@ void daliCreatePipeline(daliPipelineHandle *pipe_handle, const char *serialized_
...
     stream = dali::CUDAStream::Create(true);
   }
+  auto bs_map = std::make_unique<batch_size_map_t>();
+
   pipe_handle->ws = ws.release();
   pipe_handle->copy_stream = stream.release();
   pipe_handle->pipe = pipeline.release();
-
-  auto bs_map = std::make_unique<batch_size_map_t>();
   pipe_handle->batch_sizes_map = bs_map.release();
```
Comment (mzient) [inline dali/c_api/c_api.cc:185]:
```suggestion
  pipe_handle->batch_size_map = bs_map.release();
```

File: dali/c_api/c_api_test.cc [hunk 4]
```diff
@@ -659,25 +697,24 @@ TYPED_TEST(CApiTest, ForceNoCopyFail) {
...
   daliCreatePipeline(&handle, serialized.c_str(), serialized.size(), batch_size, num_thread,
                      device_id, false, prefetch_queue_depth, prefetch_queue_depth,
                      prefetch_queue_depth, false);
 
-  SequentialFill(view<uint8_t>(input_cpu), 42);
-  // Unnecessary copy in case of CPUBackend, makes the code generic across Backends
-  input.Copy(input_cpu, cuda_stream);
+    SequentialFill(TensorListView<StorageCPU, uint8_t>(input_cpu.get(), input_shape), 42);
+    // Unnecessary copy in case of CPUBackend, makes the code generic across Backends
+    MemCopy(input.get(), input_cpu.get(), num_elems, cuda_stream);
```
Comment (mzient) [inline dali/c_api/c_api_test.cc:707]: Hmmm... this seems to make _more_ assumptions about buffer contiguity, not less.

Comment (mzient) [inline dali/c_api/c_api_test.cc:707]: If it's unnecessary and the code is not templated, then why?

**Review by mzient — APPROVED** [review]

File: dali/c_api/c_api.cc (hunk 2 above)
Comment (klecki) [reply dali/c_api/c_api.cc:202]: done

File: dali/c_api/c_api.cc (hunk 3 above)
Comment (klecki) [reply dali/c_api/c_api.cc]: done

**Comment by klecki** [comment]
!build

**Pull request merged**